[Bug]: ColumnarScan with Reverse pathkey can violate Merge Join ordering

### What type of bug is this?

Unexpected error

### What subsystems and features are affected?

Query planner

### What happened?

# ColumnarScan with Reverse pathkey can violate Merge Join ordering

## Summary

A query that joins a compressed TimescaleDB hypertable to a reference relation on a
column participating in `compress_orderby` can fail with:

```text
ERROR: XX000: mergejoin input data is out of order
```

The error is raised by PostgreSQL's Merge Join executor (`nodeMergejoin.c`) when it
detects that an input stream advertised as sorted is not actually sorted at the
row level.

The failing plans contain `Custom Scan (ColumnarScan)` nodes with `Reverse: true`
under a `Merge Append`. The `Merge Append` advertises a sort key matching the
compressed hypertable ordering, and the parent `Merge Join` relies on that order.

## Environment

Production environment where the issue was first observed:

- PostgreSQL: `17.9` (`x86_64-pc-linux-gnu`, GCC 11.5)
- TimescaleDB: `2.26.4`
- Platform: Linux x86_64

Self-contained repro environment:

- Docker image: `timescale/timescaledb:2.26.4-pg17`
- PostgreSQL: `17.9`
- TimescaleDB: `2.26.4`
- Docker architecture depends on the host; the local verification log used
  `aarch64-unknown-linux-musl`, while the production failure is on x86_64.

## Affected areas

- Compression / compressed chunk scan
- Columnstore / Hypercore / `ColumnarScan`
- TimescaleDB planner pathkey generation
- PostgreSQL executor surface exposed through `Merge Join`

## Symptoms

- The error is deterministic for affected parameter combinations.
- `SET enable_mergejoin = off;` works around the issue.
- `SET timescaledb.enable_decompression_sorted_merge = off;` does **not** work
  around the issue.
- The failing plan includes `ColumnarScan` with `Reverse: true`.
- The child compressed-table path is sorted by metadata columns such as
  `_ts_meta_min_1`, `_ts_meta_max_1`, `_ts_meta_min_2`, `_ts_meta_max_2`, but the
  rows emitted by `ColumnarScan` are not guaranteed to preserve the corresponding
  row-level order.

## Observed plan shape

Anonymized production shape:

```text
Merge Join
  Merge Cond: (m.series_id = ref.series_id)
  ->  Merge Append
        Sort Key: m.series_id, m.ts
        ->  Custom Scan (ColumnarScan) on _hyper_X_chunk1 m_1
              Reverse: true
              ->  Sort
                    Sort Key: compress_..._ts_meta_min_1,
                              compress_..._ts_meta_max_1,
                              compress_..._ts_meta_min_2,
                              compress_..._ts_meta_max_2
                    ->  Seq Scan on compress_hyper_X_chunk1
        ->  Custom Scan (ColumnarScan) on _hyper_X_chunk2 m_2
              Reverse: true
              ->  Sort
                    Sort Key: same metadata shape
                    ->  Seq Scan on compress_hyper_X_chunk2
        ->  ...
  ->  Index Scan or Sort on ref
```

The parent `Merge Join` trusts the `Merge Append` stream to be ordered by
`(series_id, ts)`. That stream depends on each `ColumnarScan` preserving the
advertised row-level order. In affected cases, this invariant is violated.

## Workarounds tested

| Setting | Result |
| --- | --- |
| Default settings | Fails with `XX000: mergejoin input data is out of order` |
| `SET enable_mergejoin = off;` | Query succeeds |
| `SET timescaledb.enable_decompression_sorted_merge = off;` | Still fails |

This suggests the issue is not limited to the optional sorted decompression merge
path. The more general problem appears to be that `ColumnarScan` can advertise a
pathkey that is not guaranteed at the row level, especially with `Reverse: true`
and overlapping compressed-batch metadata ranges.

## Production-relevant factors

The following properties were present in the production plan or were useful in
building the repro:

- The join key (`series_id` in the repro) participates in `compress_orderby`.
- Additional filter columns are also in `compress_orderby`, so compressed chunks
  are filtered through `_ts_meta_min_N` / `_ts_meta_max_N` metadata predicates.
- The time predicate spans multiple chunks, causing a `Merge Append` over
  multiple `ColumnarScan` inputs.
- `ColumnarScan` is executed with `Reverse: true`.
- Multiple compressed batches can have overlapping orderby-key metadata ranges.
  This can happen naturally after incremental/continuous compression, where new
  compressed batches are appended to already-compressed chunks rather than being
  globally merged into a single non-overlapping batch sequence.

## Self-contained repro

The attached `docker-compose.yml` is a single-file repro. It uses only the
TimescaleDB Docker image and inline SQL configs.

Run:

```bash
cd /tmp/timescaledb-mergejoin-repro
docker compose down -v
docker compose up --abort-on-container-exit
```

Expected verdict from the current repro:

```text
VERDICT MATRIX
 case_id | label                                                | result
---------+------------------------------------------------------+------------
 A       | default settings                                     | REPRODUCED
 B       | enable_mergejoin = off                               | OK
 C       | enable_decompression_sorted_merge = off              | REPRODUCED
 D       | force merge join                                     | REPRODUCED
 E       | production plan shape (Seq Scan + Sort + Merge Join) | REPRODUCED
 F       | production shape + decompression_sorted_merge=off    | REPRODUCED

Per-case detail:
 A | XX000: mergejoin input data is out of order
 B | count=200
 C | XX000: mergejoin input data is out of order
 D | XX000: mergejoin input data is out of order
 E | XX000: mergejoin input data is out of order
 F | XX000: mergejoin input data is out of order
```

### Important note about the repro

The production failure occurs on naturally generated compressed data.

The self-contained Docker repro deliberately injects one "rogue" compressed batch
by copying an existing valid row from the underlying
`_timescaledb_internal.compress_hyper_*_chunk` table and changing only
`_ts_meta_min_1` / `_ts_meta_max_1`. The compressed payload remains valid, but
the advertised metadata pathkey becomes inconsistent with the actual row-level
values emitted by `ColumnarScan`.

This is intentional: it creates, in a small deterministic test, the same class of
planner/executor contract violation observed in production: metadata-level sort
order is trusted as row-level order.

The repro should therefore be treated as a minimal pathkey-contract test case,
not as a claim that user code modified internal TimescaleDB tables in
production.

## Why this appears TimescaleDB-specific

PostgreSQL's Merge Join is enforcing a standard executor invariant: both merge
inputs must arrive in sorted order. The failing input order is provided through a
TimescaleDB `ColumnarScan` path and its advertised pathkeys.

The Timescale-specific parts of the failing plan are:

- `Custom Scan (ColumnarScan)`
- `Reverse: true`
- compressed-table metadata sort keys such as `_ts_meta_min_N` and
  `_ts_meta_max_N`
- use of compressed-batch metadata ranges as a proxy for row-level ordering

If `ColumnarScan` cannot guarantee strict row-level ordering across compressed
batches, especially in reverse scan mode, it should not advertise a pathkey that
allows the planner to avoid an explicit row-level sort before `Merge Join`.

## Expected behavior

One of the following should happen:

1. `ColumnarScan` should only advertise pathkeys it can guarantee at the
   row-emission level.
2. If the pathkey is intended to be valid, `ColumnarScan` should enforce strict
   row-level ordering across compressed batches, including `Reverse: true`.
3. If some compressed batch layouts cannot satisfy the pathkey, the planner
   should add an explicit row-level `Sort` above the compressed scan or choose a
   non-merge join plan.

## Current user workaround

Until the extension-side pathkey contract is fixed, the reliable workaround is:

```sql
SET enable_mergejoin = off;
```

`timescaledb.enable_decompression_sorted_merge = off` is not sufficient for this
failure mode.


### TimescaleDB version affected

2.26.4

### PostgreSQL version used

17.9

### What operating system did you use?

RadHat 11.5.0-11

### What installation method did you use?

RPM

### What platform did you run on?

On prem/Self-hosted

### Relevant log output and stack trace

```bash

```

### How can we reproduce the bug?

```bash
# Self-contained reproduction for:
#   ERROR: XX000: mergejoin input data is out of order
# raised by Postgres when a Merge Join consumes rows from a
# TimescaleDB ColumnarScan whose advertised pathkey is not strictly
# honored at the row level.
#
# Single-command run:
#   docker compose up --abort-on-container-exit
#
# Re-run after edits in this file:
#   docker compose down -v && docker compose up --abort-on-container-exit
#
# The `-v` is required after editing schema/data sections, because
# Postgres only re-runs init scripts on a fresh data dir.

services:
  db:
    image: timescale/timescaledb:2.26.4-pg17
    container_name: ts-mergejoin-db
    environment:
      POSTGRES_PASSWORD: postgres
      POSTGRES_USER: postgres
      POSTGRES_DB: repro
    ports:
      - "5499:5432"
    configs:
      - source: schema_sql
        target: /docker-entrypoint-initdb.d/01-schema.sql
      - source: data_sql
        target: /docker-entrypoint-initdb.d/02-data.sql
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres -d repro"]
      interval: 3s
      timeout: 10s
      retries: 60
      start_period: 30s
    command:
      - postgres
      - -c
      - shared_preload_libraries=timescaledb
      - -c
      - max_wal_size=2GB
      - -c
      - log_min_messages=warning

  repro:
    image: timescale/timescaledb:2.26.4-pg17
    container_name: ts-mergejoin-repro
    depends_on:
      db:
        condition: service_healthy
    environment:
      PGPASSWORD: postgres
    configs:
      - source: verdict_sql
        target: /verdict.sql
    entrypoint:
      - psql
      - -h
      - db
      - -U
      - postgres
      - -d
      - repro
      - -X
      - -f
      - /verdict.sql
    restart: "no"

configs:
  schema_sql:
    content: |
      \echo
      \echo '=== Versions ==='
      SELECT version();
      SELECT extname, extversion FROM pg_extension WHERE extname = 'timescaledb';

      \echo
      \echo '=== Schema setup ==='

      CREATE EXTENSION IF NOT EXISTS timescaledb;

      DROP TABLE IF EXISTS metrics CASCADE;
      CREATE TABLE metrics (
          ts          timestamptz NOT NULL,
          series_id   bigint      NOT NULL,
          group_id    int         NOT NULL,
          batch_id    int         NOT NULL,
          value       numeric
      );

      -- Small chunks => many chunks => Merge Append over many ColumnarScans.
      SELECT create_hypertable(
          'metrics',
          'ts',
          chunk_time_interval => interval '6 hours'
      );

      -- KEY: group_id, batch_id are in compress_orderby (not segmentby).
      -- This mirrors the production EXPLAIN filter pattern:
      --   Filter: _ts_meta_min_3 <= 26 AND _ts_meta_max_3 >= 26
      --       AND _ts_meta_min_4 <= 42780 AND _ts_meta_max_4 >= 42780
      -- meaning the chunk-level filter checks min/max RANGE, not exact
      -- match. Batches whose [min, max] include 26/42780 survive even
      -- if they contain ZERO matching rows (false positives that pass
      -- the meta filter). This causes batches with overlapping series_id
      -- ranges to coexist post-filter — the precondition for the bug.
      ALTER TABLE metrics SET (
          timescaledb.compress,
          timescaledb.compress_segmentby = '',
          timescaledb.compress_orderby   = 'series_id DESC, ts, group_id DESC, batch_id DESC'
      );

      DROP TABLE IF EXISTS series_ref;
      CREATE TABLE series_ref (
          series_id bigint PRIMARY KEY,
          label     text
      );

  data_sql:
    content: |
      \echo
      \echo '=== Loading data ==='
      \echo 'Goal: many compressed batches per chunk per segmentby combo so'
      \echo 'planner is forced to insert a Sort node over _ts_meta_min/max'
      \echo 'between Seq Scan and ColumnarScan.'
      \echo
      \echo 'Layout:'
      \echo '  - 6h chunks, 12h span = 2 chunks'
      \echo '  - 500 series x 720 timestamps/chunk = 360k rows / segment'
      \echo '  - 3 (group_id, batch_id) segments per chunk'
      \echo '  - => ~360 batches/chunk/segment after filter'

      \echo 'Phase 1: 200 series x 360 ts (1min) x 1 chunk x 3 segments'
      INSERT INTO metrics (ts, series_id, group_id, batch_id, value)
      SELECT
          g.ts,
          s.series_id,
          26,
          b.batch_id,
          random() * 100
      FROM generate_series(
          '2025-06-01 00:00:00'::timestamptz,
          '2025-06-01 06:00:00'::timestamptz,
          '1 minute'
      ) g(ts)
      CROSS JOIN generate_series(1, 200) AS s(series_id)
      CROSS JOIN (VALUES (42780), (42781), (42782)) AS b(batch_id);

      \echo
      \echo '=== Phase 1 compress ==='
      SELECT count(*) AS chunks_compressed
      FROM (
          SELECT compress_chunk(c) FROM show_chunks('metrics') c
      ) s;

      \echo
      \echo 'Phase 2: half-size insert with DIFFERENT series_ids range to'
      \echo 'create overlapping batches when recompressed (key trigger).'
      INSERT INTO metrics (ts, series_id, group_id, batch_id, value)
      SELECT
          g.ts,
          s.series_id,
          26,
          b.batch_id,
          random() * 100
      FROM generate_series(
          '2025-06-01 00:00:00'::timestamptz,
          '2025-06-01 06:00:00'::timestamptz,
          '1 minute'
      ) g(ts)
      CROSS JOIN generate_series(50, 150) AS s(series_id)  -- overlapping range
      CROSS JOIN (VALUES (42780)) AS b(batch_id);

      \echo
      \echo '=== Phase 2 compress (NO recompress: append-only batches) ==='
      \echo 'Without recompress=true, TS only compresses the staging area'
      \echo '(rows added after initial compression) into NEW batches. These'
      \echo 'batches are APPENDED — their min/max ranges OVERLAP with the'
      \echo 'phase-1 batches, which is the precondition for the bug.'
      DO $$$$
      DECLARE c regclass;
      BEGIN
          FOR c IN SELECT show_chunks('metrics')
          LOOP
              BEGIN
                  PERFORM compress_chunk(c);
              EXCEPTION WHEN OTHERS THEN
                  RAISE NOTICE 'compress on %: %', c, SQLERRM;
                  -- Fallback: use recompress if append-only not supported.
                  BEGIN
                      PERFORM compress_chunk(c, recompress => true);
                  EXCEPTION WHEN OTHERS THEN
                      RAISE NOTICE 'recompress on %: %', c, SQLERRM;
                  END;
              END;
          END LOOP;
      END$$$$;

      \echo
      \echo '=== Injecting rogue compressed batch metadata ==='
      \echo 'Copy one valid compressed row but lie about _ts_meta_min_1/_ts_meta_max_1.'
      \echo 'The compressed payload stays valid; only the advertised pathkey metadata'
      \echo 'is made inconsistent with the row-level series_id values.'
      DO $$$$
      DECLARE
          target_rel regclass;
          inserted_rows bigint;
      BEGIN
          SELECT relid
          INTO target_rel
          FROM (
              SELECT (n.nspname || '.' || c.relname)::regclass AS relid,
                     c.reltuples
              FROM pg_class c
              JOIN pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = '_timescaledb_internal'
                AND c.relname LIKE 'compress_hyper_%_chunk'
              ORDER BY c.reltuples DESC NULLS LAST, c.relname
              LIMIT 1
          ) s;

          IF target_rel IS NULL THEN
              RAISE NOTICE 'No compressed chunk found for rogue metadata injection';
              RETURN;
          END IF;

          EXECUTE format(
              $$q$$
              INSERT INTO %s (
                  _ts_meta_count,
                  _ts_meta_min_2,
                  _ts_meta_max_2,
                  ts,
                  _ts_meta_min_1,
                  _ts_meta_max_1,
                  series_id,
                  _ts_meta_min_3,
                  _ts_meta_max_3,
                  group_id,
                  _ts_meta_min_4,
                  _ts_meta_max_4,
                  batch_id,
                  value
              )
              SELECT
                  _ts_meta_count,
                  _ts_meta_min_2,
                  _ts_meta_max_2,
                  ts,
                  1::bigint AS _ts_meta_min_1,
                  1::bigint AS _ts_meta_max_1,
                  series_id,
                  _ts_meta_min_3,
                  _ts_meta_max_3,
                  group_id,
                  _ts_meta_min_4,
                  _ts_meta_max_4,
                  batch_id,
                  value
              FROM %s
              WHERE _ts_meta_min_1 >= 150
                AND _ts_meta_min_3 <= 26
                AND _ts_meta_max_3 >= 26
                AND _ts_meta_min_4 <= 42780
                AND _ts_meta_max_4 >= 42780
              ORDER BY _ts_meta_min_1 DESC, _ts_meta_max_1 DESC
              LIMIT 1
              $$q$$,
              target_rel,
              target_rel
          );

          GET DIAGNOSTICS inserted_rows = ROW_COUNT;
          RAISE NOTICE 'Rogue metadata batch insert into %: rows=%', target_rel, inserted_rows;
      END$$$$;

      \echo
      \echo '=== Dropping auto-created indexes on compressed chunks ==='
      \echo 'TimescaleDB creates a composite index (segmentby..., _ts_meta_min_1)'
      \echo 'on each compressed chunk. With it present, the planner picks an'
      \echo 'Index Scan that already delivers batches in the right order.'
      \echo 'Production EXPLAIN shows Seq Scan + Sort over _ts_meta_min/max'
      \echo '— suggesting the index is not used (or absent). Drop indexes here'
      \echo 'to force Postgres into the same plan shape.'
      DO $$$$
      DECLARE
          r record;
          dropped int := 0;
      BEGIN
          FOR r IN
              SELECT (ns.nspname || '.' || c.relname)::text AS idx
                FROM pg_index i
                JOIN pg_class c   ON c.oid  = i.indexrelid
                JOIN pg_namespace ns ON ns.oid = c.relnamespace
                JOIN pg_class t   ON t.oid  = i.indrelid
               WHERE ns.nspname = '_timescaledb_internal'
                 AND t.relname LIKE 'compress_hyper_%_chunk'
                 AND c.relname LIKE '%_ts_meta_min_1%idx'
          LOOP
              EXECUTE 'DROP INDEX ' || r.idx;
              dropped := dropped + 1;
          END LOOP;
          RAISE NOTICE 'Dropped % compress indexes', dropped;
      END$$$$;

      INSERT INTO series_ref
          SELECT s, 'L' || s FROM generate_series(1, 500) s;

      ANALYZE metrics;
      ANALYZE series_ref;

      \echo
      \echo '=== Storage layout ==='
      SELECT
          chunk_schema,
          chunk_name,
          is_compressed,
          pg_size_pretty(
              pg_total_relation_size(format('%I.%I', chunk_schema, chunk_name)::regclass)
          ) AS total
      FROM timescaledb_information.chunks
      WHERE hypertable_name = 'metrics'
      ORDER BY chunk_name;

      \echo
      \echo '=== Timescale compression API / GUC diagnostics ==='
      SELECT
          n.nspname AS schema,
          p.proname AS function,
          pg_get_function_identity_arguments(p.oid) AS args
      FROM pg_proc p
      JOIN pg_namespace n ON n.oid = p.pronamespace
      WHERE n.nspname IN ('public', '_timescaledb_internal')
        AND (
            p.proname ILIKE '%compress%'
            OR p.proname ILIKE '%recompress%'
            OR p.proname ILIKE '%columnar%'
            OR p.proname ILIKE '%hypercore%'
        )
      ORDER BY schema, function, args;

      SELECT name, setting, vartype, context
      FROM pg_settings
      WHERE name LIKE 'timescaledb.%'
        AND (
            name LIKE '%compress%'
            OR name LIKE '%decompress%'
            OR name LIKE '%columnar%'
            OR name LIKE '%hypercore%'
            OR name LIKE '%vector%'
        )
      ORDER BY name;

      \echo
      \echo '=== Compressed table column layout ==='
      SELECT
          c.relname AS compressed_table,
          a.attnum,
          a.attname,
          format_type(a.atttypid, a.atttypmod) AS type
      FROM pg_class c
      JOIN pg_namespace n ON n.oid = c.relnamespace
      JOIN pg_attribute a ON a.attrelid = c.oid
      WHERE n.nspname = '_timescaledb_internal'
        AND c.relname LIKE 'compress_hyper_%_chunk'
        AND a.attnum > 0
        AND NOT a.attisdropped
      ORDER BY c.relname, a.attnum;

      \echo
      \echo '=== Compressed batch range diagnostics ==='
      DO $$$$
      DECLARE
          r record;
          total_batches bigint;
          min_series bigint;
          max_series bigint;
          duplicate_min_values bigint;
          adjacent_overlaps bigint;
          matching_meta_batches bigint;
      BEGIN
          FOR r IN
              SELECT (n.nspname || '.' || c.relname)::regclass AS relid,
                     c.relname AS relname
              FROM pg_class c
              JOIN pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = '_timescaledb_internal'
                AND c.relname LIKE 'compress_hyper_%_chunk'
              ORDER BY c.relname
          LOOP
              EXECUTE format(
                  $$q$$
                  SELECT
                      count(*)::bigint,
                      min(_ts_meta_min_1)::bigint,
                      max(_ts_meta_max_1)::bigint,
                      count(*) FILTER (WHERE dup_count > 1)::bigint,
                      count(*) FILTER (WHERE prev_max IS NOT NULL AND _ts_meta_min_1 <= prev_max)::bigint,
                      count(*) FILTER (
                          WHERE _ts_meta_min_3 <= 26
                            AND _ts_meta_max_3 >= 26
                            AND _ts_meta_min_4 <= 42780
                            AND _ts_meta_max_4 >= 42780
                      )::bigint
                  FROM (
                      SELECT
                          _ts_meta_min_1,
                          _ts_meta_max_1,
                          _ts_meta_min_3,
                          _ts_meta_max_3,
                          _ts_meta_min_4,
                          _ts_meta_max_4,
                          count(*) OVER (PARTITION BY _ts_meta_min_1) AS dup_count,
                          lag(_ts_meta_max_1) OVER (
                              ORDER BY _ts_meta_min_1, _ts_meta_max_1, _ts_meta_min_2 DESC, _ts_meta_max_2 DESC
                          ) AS prev_max
                      FROM %s
                  ) s
                  $$q$$,
                  r.relid
              )
              INTO total_batches, min_series, max_series, duplicate_min_values,
                   adjacent_overlaps, matching_meta_batches;

              RAISE NOTICE
                  '%: total_batches=%, matching_meta_batches=%, series_range=[%,%], duplicate_min_rows=%, adjacent_overlaps=%',
                  r.relname,
                  total_batches,
                  matching_meta_batches,
                  min_series,
                  max_series,
                  duplicate_min_values,
                  adjacent_overlaps;
          END LOOP;
      END$$$$;

  verdict_sql:
    content: |
      \set ON_ERROR_STOP off
      \pset border 1
      \pset format aligned

      \echo
      \echo === Versions ===
      SELECT version();
      SELECT extversion FROM pg_extension WHERE extname = 'timescaledb';

      \echo
      \echo === EXPLAIN (default settings) ===
      \echo Query mirrors production: filter_aps CTE joined to compressed
      \echo hypertable, with WindowAgg + rn=1 filter on top. This shape
      \echo lets the planner consider a Merge Join over ColumnarScan output.
      EXPLAIN (VERBOSE, SETTINGS)
      WITH filter_aps AS NOT MATERIALIZED (
          SELECT series_id, label
            FROM series_ref
           WHERE series_id BETWEEN 1 AND 200
      )
      , ts AS (
          SELECT
              m.ts, m.series_id, m.value, fa.label,
              row_number() OVER (
                  PARTITION BY m.series_id
                  ORDER BY m.ts DESC
              ) AS rn
          FROM metrics m
          JOIN filter_aps fa ON fa.series_id = m.series_id
          WHERE m.ts >= '2025-06-01'::timestamptz
            AND m.ts <  '2025-08-01'::timestamptz
            AND m.group_id = 26
            AND m.batch_id = 42780
      )
      SELECT * FROM ts WHERE rn = 1;

      \echo
      \echo === EXPLAIN with case E settings (what does Seq Scan path look like?) ===
      BEGIN;
      SET LOCAL enable_indexscan = off;
      SET LOCAL enable_bitmapscan = off;
      SET LOCAL enable_hashjoin = off;
      SET LOCAL enable_nestloop = off;
      SET LOCAL work_mem = '64kB';
      EXPLAIN (VERBOSE)
      WITH filter_aps AS NOT MATERIALIZED (
          SELECT series_id, label
            FROM series_ref
           WHERE series_id BETWEEN 1 AND 200
      )
      , ts AS (
          SELECT
              m.ts, m.series_id, m.value, fa.label,
              row_number() OVER (
                  PARTITION BY m.series_id
                  ORDER BY m.ts DESC
              ) AS rn
          FROM metrics m
          JOIN filter_aps fa ON fa.series_id = m.series_id
          WHERE m.ts >= '2025-06-01'::timestamptz
            AND m.ts <  '2025-08-01'::timestamptz
            AND m.group_id = 26
            AND m.batch_id = 42780
      )
      SELECT * FROM ts WHERE rn = 1;
      ROLLBACK;

      \echo
      \echo === Running test cases ===

      DROP TABLE IF EXISTS verdict;
      CREATE TEMP TABLE verdict (
          case_id  text,
          label    text,
          settings text,
          result   text,
          detail   text
      );

      DO $$do$$
      DECLARE
          v_query constant text := $$q$$
              -- Mirrors production: a small reference CTE joined to the
              -- compressed hypertable on the orderby column, with a
              -- WindowAgg filter on top. This shape makes the planner
              -- attempt a Merge Join between the compressed-side stream
              -- (Merge Append over ColumnarScans) and the reference set.
              WITH filter_aps AS NOT MATERIALIZED (
                  SELECT series_id, label
                    FROM series_ref
                   WHERE series_id BETWEEN 1 AND 200
              )
              , ts AS (
                  SELECT
                      m.ts, m.series_id, m.value, fa.label,
                      row_number() OVER (
                          PARTITION BY m.series_id
                          ORDER BY m.ts DESC
                      ) AS rn
                  FROM metrics m
                  JOIN filter_aps fa ON fa.series_id = m.series_id
                  WHERE m.ts >= '2025-06-01'::timestamptz
                    AND m.ts <  '2025-08-01'::timestamptz
                    AND m.group_id = 26
                    AND m.batch_id = 42780
              )
              SELECT count(*) FROM ts WHERE rn = 1
          $$q$$;
          cnt bigint;
      BEGIN
          BEGIN
              EXECUTE v_query INTO cnt;
              INSERT INTO verdict VALUES (
                  'A', 'default settings', '<none>', 'OK', 'count=' || cnt);
          EXCEPTION WHEN OTHERS THEN
              INSERT INTO verdict VALUES (
                  'A', 'default settings', '<none>',
                  CASE WHEN SQLERRM LIKE '%mergejoin input data is out of order%'
                       THEN 'REPRODUCED' ELSE 'OTHER_ERROR' END,
                  SQLSTATE || ': ' || SQLERRM);
          END;

          BEGIN
              PERFORM set_config('enable_mergejoin', 'on', true);
              PERFORM set_config('enable_hashjoin', 'on', true);
              PERFORM set_config('enable_nestloop', 'on', true);
              PERFORM set_config('enable_indexscan', 'on', true);
              PERFORM set_config('enable_bitmapscan', 'on', true);
              PERFORM set_config('timescaledb.enable_decompression_sorted_merge', 'on', true);
              PERFORM set_config('work_mem', '4MB', true);
              SET LOCAL enable_mergejoin = off;
              EXECUTE v_query INTO cnt;
              INSERT INTO verdict VALUES (
                  'B', 'enable_mergejoin = off',
                  'enable_mergejoin=off', 'OK', 'count=' || cnt);
          EXCEPTION WHEN OTHERS THEN
              INSERT INTO verdict VALUES (
                  'B', 'enable_mergejoin = off',
                  'enable_mergejoin=off',
                  CASE WHEN SQLERRM LIKE '%mergejoin input data is out of order%'
                       THEN 'REPRODUCED' ELSE 'OTHER_ERROR' END,
                  SQLSTATE || ': ' || SQLERRM);
          END;

          BEGIN
              PERFORM set_config('enable_mergejoin', 'on', true);
              PERFORM set_config('enable_hashjoin', 'on', true);
              PERFORM set_config('enable_nestloop', 'on', true);
              PERFORM set_config('enable_indexscan', 'on', true);
              PERFORM set_config('enable_bitmapscan', 'on', true);
              PERFORM set_config('timescaledb.enable_decompression_sorted_merge', 'on', true);
              PERFORM set_config('work_mem', '4MB', true);
              SET LOCAL timescaledb.enable_decompression_sorted_merge = off;
              EXECUTE v_query INTO cnt;
              INSERT INTO verdict VALUES (
                  'C', 'enable_decompression_sorted_merge = off',
                  'ts.enable_decompression_sorted_merge=off', 'OK', 'count=' || cnt);
          EXCEPTION WHEN OTHERS THEN
              INSERT INTO verdict VALUES (
                  'C', 'enable_decompression_sorted_merge = off',
                  'ts.enable_decompression_sorted_merge=off',
                  CASE WHEN SQLERRM LIKE '%mergejoin input data is out of order%'
                       THEN 'REPRODUCED' ELSE 'OTHER_ERROR' END,
                  SQLSTATE || ': ' || SQLERRM);
          END;

          BEGIN
              PERFORM set_config('enable_mergejoin', 'on', true);
              PERFORM set_config('enable_hashjoin', 'on', true);
              PERFORM set_config('enable_nestloop', 'on', true);
              PERFORM set_config('enable_indexscan', 'on', true);
              PERFORM set_config('enable_bitmapscan', 'on', true);
              PERFORM set_config('timescaledb.enable_decompression_sorted_merge', 'on', true);
              PERFORM set_config('work_mem', '4MB', true);
              SET LOCAL enable_hashjoin = off;
              SET LOCAL enable_nestloop = off;
              EXECUTE v_query INTO cnt;
              INSERT INTO verdict VALUES (
                  'D', 'force merge join',
                  'enable_hashjoin=off, enable_nestloop=off', 'OK', 'count=' || cnt);
          EXCEPTION WHEN OTHERS THEN
              INSERT INTO verdict VALUES (
                  'D', 'force merge join',
                  'enable_hashjoin=off, enable_nestloop=off',
                  CASE WHEN SQLERRM LIKE '%mergejoin input data is out of order%'
                       THEN 'REPRODUCED' ELSE 'OTHER_ERROR' END,
                  SQLSTATE || ': ' || SQLERRM);
          END;

          -- Case E: force Seq Scan + Sort path on compressed chunks
          -- AND make top-level Sort prohibitively expensive (low work_mem)
          -- so the planner chooses Merge Append + asks ColumnarScan for
          -- ordered output (which then needs the inner Sort over
          -- _ts_meta_min/max — the production trigger shape).
          BEGIN
              PERFORM set_config('enable_mergejoin', 'on', true);
              PERFORM set_config('enable_hashjoin', 'on', true);
              PERFORM set_config('enable_nestloop', 'on', true);
              PERFORM set_config('enable_indexscan', 'on', true);
              PERFORM set_config('enable_bitmapscan', 'on', true);
              PERFORM set_config('timescaledb.enable_decompression_sorted_merge', 'on', true);
              PERFORM set_config('work_mem', '4MB', true);
              SET LOCAL enable_indexscan = off;
              SET LOCAL enable_bitmapscan = off;
              SET LOCAL enable_hashjoin = off;
              SET LOCAL enable_nestloop = off;
              SET LOCAL work_mem = '64kB';
              EXECUTE v_query INTO cnt;
              INSERT INTO verdict VALUES (
                  'E', 'production plan shape (Seq Scan + Sort + Merge Join)',
                  'idx/bitmap/hash/nestloop=off, work_mem=64kB',
                  'OK', 'count=' || cnt);
          EXCEPTION WHEN OTHERS THEN
              INSERT INTO verdict VALUES (
                  'E', 'production plan shape (Seq Scan + Sort + Merge Join)',
                  'idx/bitmap/hash/nestloop=off, work_mem=64kB',
                  CASE WHEN SQLERRM LIKE '%mergejoin input data is out of order%'
                       THEN 'REPRODUCED' ELSE 'OTHER_ERROR' END,
                  SQLSTATE || ': ' || SQLERRM);
          END;

          -- Case F: case E + disable TS decompression_sorted_merge.
          -- If TS still does internal merging across overlapping batches,
          -- turning it off should expose the row-order violation.
          BEGIN
              PERFORM set_config('enable_mergejoin', 'on', true);
              PERFORM set_config('enable_hashjoin', 'on', true);
              PERFORM set_config('enable_nestloop', 'on', true);
              PERFORM set_config('enable_indexscan', 'on', true);
              PERFORM set_config('enable_bitmapscan', 'on', true);
              PERFORM set_config('timescaledb.enable_decompression_sorted_merge', 'on', true);
              PERFORM set_config('work_mem', '4MB', true);
              SET LOCAL enable_indexscan = off;
              SET LOCAL enable_bitmapscan = off;
              SET LOCAL enable_hashjoin = off;
              SET LOCAL enable_nestloop = off;
              SET LOCAL work_mem = '64kB';
              SET LOCAL timescaledb.enable_decompression_sorted_merge = off;
              EXECUTE v_query INTO cnt;
              INSERT INTO verdict VALUES (
                  'F', 'production shape + decompression_sorted_merge=off',
                  'case E + ts.enable_decompression_sorted_merge=off',
                  'OK', 'count=' || cnt);
          EXCEPTION WHEN OTHERS THEN
              INSERT INTO verdict VALUES (
                  'F', 'production shape + decompression_sorted_merge=off',
                  'case E + ts.enable_decompression_sorted_merge=off',
                  CASE WHEN SQLERRM LIKE '%mergejoin input data is out of order%'
                       THEN 'REPRODUCED' ELSE 'OTHER_ERROR' END,
                  SQLSTATE || ': ' || SQLERRM);
          END;
      END
      $$do$$;

      \echo
      \echo ============================================================
      \echo                      VERDICT MATRIX
      \echo ============================================================
      SELECT case_id, label, settings, result FROM verdict ORDER BY case_id;

      \echo
      \echo Per-case detail:
      SELECT case_id, detail FROM verdict ORDER BY case_id;

      \echo
      \echo Expected with rogue compressed metadata injection:
      \echo   A default settings                                -> REPRODUCED
      \echo   B enable_mergejoin = off                          -> OK
      \echo   C enable_decompression_sorted_merge = off         -> REPRODUCED
      \echo   D force merge join                                -> REPRODUCED
      \echo   E production plan shape (Seq Scan + Sort + MJ)    -> REPRODUCED <<< the trigger
      \echo   F case E + decompression_sorted_merge = off       -> REPRODUCED
```

--- a/.unreleased/pr_9733
+++ b/.unreleased/pr_9733
@@ -1,0 +1,2 @@
+Fixes: #9731 Avoid creating overlapping batches during recompression for multi orderby configurations
+Thanks: @h0rn3t for reporting issue with recompression creating overlapping batches

--- a/src/debug_point.c
+++ b/src/debug_point.c
@@ -325,7 +325,9 @@ ts_debug_point_raise_error_oneshot(const char *name)
 		case LOCKACQUIRE_OK:
 			LockRelease(&point.tag, ShareLock, true);
 			if (LockHeldByMeCompat(&point.tag, ExclusiveLock, false))
+			{
 				break;
+			}
 			return;
 		case LOCKACQUIRE_ALREADY_HELD:
 		case LOCKACQUIRE_ALREADY_CLEAR:

--- a/tsl/src/compression/recompress.c
+++ b/tsl/src/compression/recompress.c
@@ -1059,19 +1059,26 @@ static enum Batch_match_result
 match_tuple_batch(TupleTableSlot *compressed_slot, int num_orderby, ScanKey orderby_scankeys,
 				  bool *nulls_first)
 {
-	ScanKey key;
-	for (int i = 0; i < num_orderby; i++)
+	/*
+	 * Only the leading orderby column gives a sound before/after verdict from
+	 * batch metadata. The min/max for later orderby columns are aggregated
+	 * over all rows in the batch, not conditional on the leading column, so a
+	 * tuple whose leading column is in range but whose later column is out of
+	 * range is interleaved with the batch in multi-column sort order — not
+	 * strictly before or after it.
+	 */
+	if (num_orderby >= 1)
 	{
-		key = &orderby_scankeys[i * 2];
+		ScanKey key = &orderby_scankeys[0];
 		if (!slot_key_test(compressed_slot, key))
 		{
-			return handle_null_scan(key->sk_flags, nulls_first[i], Tuple_before);
+			return handle_null_scan(key->sk_flags, nulls_first[0], Tuple_before);
 		}
 
-		key = &orderby_scankeys[i * 2 + 1];
+		key = &orderby_scankeys[1];
 		if (!slot_key_test(compressed_slot, key))
 		{
-			return handle_null_scan(key->sk_flags, nulls_first[i], Tuple_after);
+			return handle_null_scan(key->sk_flags, nulls_first[0], Tuple_after);
 		}
 	}
 

--- a/tsl/test/expected/recompress_chunk_segmentwise.out
+++ b/tsl/test/expected/recompress_chunk_segmentwise.out
@@ -763,3 +763,83 @@ EXPLAIN (COSTS OFF) SELECT a, time FROM segwise_unordered WHERE a = 2 ORDER BY a
 
 RESET timescaledb.debug_require_batch_sorted_merge;
 DROP TABLE segwise_unordered;
+----------------------------------------------------------------------------
+-- Issue #9731: segmentwise recompression must not write a new batch whose
+-- leading-orderby range is contained inside an existing batch's range.
+-- The matcher previously evaluated each orderby column independently and
+-- could return Tuple_after based on a non-leading column even when the
+-- leading column sat inside the existing batch's range. The result was a
+-- new batch shifted up inside the original, which the planner's
+-- use_compressed_sort path then concatenated out of order.
+----------------------------------------------------------------------------
+CREATE TABLE segwise_no_overlap (
+    ts        timestamptz NOT NULL,
+    series_id bigint      NOT NULL,
+    value     int
+);
+SELECT create_hypertable('segwise_no_overlap', 'ts', chunk_time_interval => interval '30 day');
+        create_hypertable         
+----------------------------------
+ (27,public,segwise_no_overlap,t)
+
+ALTER TABLE segwise_no_overlap SET (
+    timescaledb.compress,
+    timescaledb.compress_segmentby = '',
+    timescaledb.compress_orderby   = 'series_id DESC, ts ASC'
+);
+-- Initial compression: 1000 rows for series_id 1..100, ts 00:01..00:10.
+-- Produces a single batch with series_id range [1,100].
+INSERT INTO segwise_no_overlap
+SELECT '2026-05-01 00:00:00+00'::timestamptz + (t * interval '1 second'), s, s * 100 + t
+FROM generate_series(1, 100) s
+CROSS JOIN generate_series(1, 10) t;
+SELECT compress_chunk(c) FROM show_chunks('segwise_no_overlap') c;
+              compress_chunk              
+------------------------------------------
+ _timescaledb_internal._hyper_27_29_chunk
+
+-- Add rows for series_id = 50 (inside the existing batch's series_id range)
+-- with ts outside the existing batch's ts range. Pre-fix, the matcher would
+-- decide Tuple_after on ts and write a new batch [50,50] inside [1,100].
+INSERT INTO segwise_no_overlap
+SELECT '2026-05-01 06:00:00+00'::timestamptz + (t * interval '1 second'), 50, 99000 + t
+FROM generate_series(1, 500) t;
+-- Triggers segmentwise recompression (chunk is partial).
+SELECT compress_chunk(c) FROM show_chunks('segwise_no_overlap') c;
+              compress_chunk              
+------------------------------------------
+ _timescaledb_internal._hyper_27_29_chunk
+
+-- Locate the compressed chunk for the metadata check.
+SELECT cc.schema_name || '.' || cc.table_name AS comp_table_segwise
+FROM _timescaledb_catalog.chunk uc
+JOIN _timescaledb_catalog.chunk cc      ON cc.id = uc.compressed_chunk_id
+JOIN _timescaledb_catalog.hypertable h  ON h.id  = uc.hypertable_id
+WHERE h.table_name = 'segwise_no_overlap' \gset
+-- Resulting batches: must be at most adjacent on series_id, never
+-- shifted-up contained. Consecutive pairs in (min DESC, max DESC) order
+-- must satisfy prev_min >= curr_max.
+SELECT count(*) AS overlap_violations FROM (
+    SELECT _ts_meta_max_1 AS curr_hi,
+           lag(_ts_meta_min_1) OVER (
+               ORDER BY _ts_meta_min_1 DESC, _ts_meta_max_1 DESC
+           ) AS prev_lo
+    FROM :comp_table_segwise
+) t WHERE prev_lo IS NOT NULL AND prev_lo < curr_hi;
+ overlap_violations 
+--------------------
+                  0
+
+-- Row-level stream must be monotone DESC on series_id.
+SELECT count(*) AS desc_violations FROM (
+    SELECT series_id, lag(series_id) OVER () AS prev
+    FROM (
+        SELECT series_id FROM segwise_no_overlap
+        ORDER BY series_id DESC, ts ASC
+    ) ordered
+) lagged WHERE prev IS NOT NULL AND prev < series_id;
+ desc_violations 
+-----------------
+               0
+
+DROP TABLE segwise_no_overlap CASCADE;

--- a/tsl/test/sql/recompress_chunk_segmentwise.sql
+++ b/tsl/test/sql/recompress_chunk_segmentwise.sql
@@ -447,3 +447,73 @@ EXPLAIN (COSTS OFF) SELECT a, time FROM segwise_unordered WHERE a = 2 ORDER BY a
 RESET timescaledb.debug_require_batch_sorted_merge;
 
 DROP TABLE segwise_unordered;
+
+----------------------------------------------------------------------------
+-- Issue #9731: segmentwise recompression must not write a new batch whose
+-- leading-orderby range is contained inside an existing batch's range.
+-- The matcher previously evaluated each orderby column independently and
+-- could return Tuple_after based on a non-leading column even when the
+-- leading column sat inside the existing batch's range. The result was a
+-- new batch shifted up inside the original, which the planner's
+-- use_compressed_sort path then concatenated out of order.
+----------------------------------------------------------------------------
+CREATE TABLE segwise_no_overlap (
+    ts        timestamptz NOT NULL,
+    series_id bigint      NOT NULL,
+    value     int
+);
+SELECT create_hypertable('segwise_no_overlap', 'ts', chunk_time_interval => interval '30 day');
+
+ALTER TABLE segwise_no_overlap SET (
+    timescaledb.compress,
+    timescaledb.compress_segmentby = '',
+    timescaledb.compress_orderby   = 'series_id DESC, ts ASC'
+);
+
+-- Initial compression: 1000 rows for series_id 1..100, ts 00:01..00:10.
+-- Produces a single batch with series_id range [1,100].
+INSERT INTO segwise_no_overlap
+SELECT '2026-05-01 00:00:00+00'::timestamptz + (t * interval '1 second'), s, s * 100 + t
+FROM generate_series(1, 100) s
+CROSS JOIN generate_series(1, 10) t;
+
+SELECT compress_chunk(c) FROM show_chunks('segwise_no_overlap') c;
+
+-- Add rows for series_id = 50 (inside the existing batch's series_id range)
+-- with ts outside the existing batch's ts range. Pre-fix, the matcher would
+-- decide Tuple_after on ts and write a new batch [50,50] inside [1,100].
+INSERT INTO segwise_no_overlap
+SELECT '2026-05-01 06:00:00+00'::timestamptz + (t * interval '1 second'), 50, 99000 + t
+FROM generate_series(1, 500) t;
+
+-- Triggers segmentwise recompression (chunk is partial).
+SELECT compress_chunk(c) FROM show_chunks('segwise_no_overlap') c;
+
+-- Locate the compressed chunk for the metadata check.
+SELECT cc.schema_name || '.' || cc.table_name AS comp_table_segwise
+FROM _timescaledb_catalog.chunk uc
+JOIN _timescaledb_catalog.chunk cc      ON cc.id = uc.compressed_chunk_id
+JOIN _timescaledb_catalog.hypertable h  ON h.id  = uc.hypertable_id
+WHERE h.table_name = 'segwise_no_overlap' \gset
+
+-- Resulting batches: must be at most adjacent on series_id, never
+-- shifted-up contained. Consecutive pairs in (min DESC, max DESC) order
+-- must satisfy prev_min >= curr_max.
+SELECT count(*) AS overlap_violations FROM (
+    SELECT _ts_meta_max_1 AS curr_hi,
+           lag(_ts_meta_min_1) OVER (
+               ORDER BY _ts_meta_min_1 DESC, _ts_meta_max_1 DESC
+           ) AS prev_lo
+    FROM :comp_table_segwise
+) t WHERE prev_lo IS NOT NULL AND prev_lo < curr_hi;
+
+-- Row-level stream must be monotone DESC on series_id.
+SELECT count(*) AS desc_violations FROM (
+    SELECT series_id, lag(series_id) OVER () AS prev
+    FROM (
+        SELECT series_id FROM segwise_no_overlap
+        ORDER BY series_id DESC, ts ASC
+    ) ordered
+) lagged WHERE prev IS NOT NULL AND prev < series_id;
+
+DROP TABLE segwise_no_overlap CASCADE;

